### DISCUSSION TITLE
Possible to exceeded the size of 256 alignment in uniform buffer offset ?

### DISCUSSION THREAD

**Discussion by jay19240** [gpuweb/gpuweb#4267] [Q&A] [answered]
Please, what is the correct offset for a big uniform variable that exceeded 256 bytes ?

Work only with 256 blocks ?
`const correctAlignment = Math.ceil(data.byteLength / 256);`

Or with any size bigger than 256 bytes ?
`const correctAlignment = data.byteLength; // ex: 960`

**Comment by kainino0x** [comment]
Can you provide more context to your question? I'm not sure what offsets you're talking about or what the setup of your uniform buffer is.

**Reply by jay19240** [reply]
Offset present in GPUBufferBinding, when you call createBindGroup(descriptor)

**Comment by jay19240** [comment]
The context is GPUBufferBinding, when you call createBindGroup(descriptor)
```ts
allocate(size: number = 1): void {    
    this.bindGroups = [];
    this.buffer.destroy();
    this.buffer = this.device.createBuffer({ size: size * this.inputs.length * MIN_UNIFORM_BUFFER_OFFSET_ALIGNMENT, usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST });

    let offset = 0;

    for (let i = 0; i < size; i++) {
      const entries: Array<GPUBindGroupEntry> = [];

      for (const input of this.inputs) {
        entries[input.binding] = { binding: input.binding, resource: { buffer: this.buffer, offset: offset, size: input.size } };
        offset += MIN_UNIFORM_BUFFER_OFFSET_ALIGNMENT; // !! here the minimum uniform buffer offset is too small for my DECALS uniform
      }

      this.bindGroups.push(this.device.createBindGroup({ layout: this.pipeline.getBindGroupLayout(this.groupIndex), entries: entries }));
    }

    this.size = size;
  }
```
The shader uniform:
```
@group(0) @binding(5) var<uniform> DECAL_COUNT: u32;
@group(0) @binding(6) var<uniform> DECALS: array<Decal, ${MAX_DECALS}>; // MAX_DECALS = 16
```
My data is bigger than 256 bytes for one uniform.
My question is therefore about what offset value should I put in this case.

**Reply by jay19240** [reply, marked as answer]
I answer myself and also for those who ask the same question.
If your uniform variable is bigger than 256, you need the following code to compute the correct offset: 
`offset = Math.ceil(data.byteLength / 256) * MIN_UNIFORM_BUFFER_OFFSET_ALIGNMENT;`

Otherwise you getting this error message:
`Offset (1344) does not satisfy the minimum BufferBindingType::Uniform alignment (256).`

**Reply by kainino0x** [reply]
Glad you worked it out! Sounds like the issue is particular to the pattern in your code where you're using a single buffer for multiple uniforms. Usually in this case you should use just *one* uniform, which has a struct type:

```wgsl
struct DecalArray {
    count: u32,
    // note: there may be padding between these members, if the alignment of the Decal type is >4 bytes.
    decals: array<Decal, ${MAX_DECALS}>,
}
@group(0) @binding(5) var<uniform> DECALS: DecalArray;
```

In WebGL, you have to use multiple uniforms for multiple pieces of data. In contranst, in WebGPU, every uniform is a buffer, so you should take advantage of this to have fewer bindings.

**Comment by jay19240** [comment]
You are right, usage of struct in this case can reduce the number of writeBuffer but in my case these uniforms are send just one time by frame so it is not really critical, i think.

**Reply by kainino0x** [reply]
It would also save on the number of separate uniform bindings into the shader. I can't say whether that would have a performance impact but there's a possibility that there's some draw-time cost to the extra bindings on some hardware/backends.

**Reply by jay19240** [reply]
Ok, it is good to known !
I suppose you are very busy but if you want collaborate on my engine (Haiku), this could be awesome !
Good bye.
